pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,523 @@
+	exec := utilexec.New()
+	minVersion, err := semver.NewVersion(IPTABLES_MIN_VERSION)
+	if err != nil {
+		return false, err
+	}
+	versionString, err := utiliptables.GetIptablesVersionString(exec)
+	if err != nil {
+		return false, err
+	}
+	version, err := semver.NewVersion(versionString)
+	if err != nil {
+		return false, err
+	}
+	return !version.LessThan(*minVersion), nil
+}
+
+type portal struct {
+	ip       net.IP
+	port     int
+	protocol api.Protocol
+}
+
+// internal struct for string service information
+type serviceInfo struct {
+	portal              portal
+	loadBalancerStatus  api.LoadBalancerStatus
+	sessionAffinityType api.ServiceAffinity
+	stickyMaxAgeMinutes int
+	endpoints           []string
+	// Deprecated, but required for back-compat (including e2e)
+	deprecatedPublicIPs []string
+}
+
+// Proxier is an iptables based proxy for connections between a localhost:lport
+// and services that provide the actual backends.
+type Proxier struct {
+	mu         sync.Mutex // protects serviceMap
+	serviceMap map[proxy.ServicePortName]*serviceInfo
+	iptables   utiliptables.Interface
+}
+
+// Proxier implements ProxyProvider
+var _ proxy.ProxyProvider = &Proxier{}
+
+// NewProxier returns a new Proxier given an iptables Interface instance.
+// Because of the iptables logic, It is assumed that there is only a single Proxier active on a machine.
+// An error will be returned if iptables fails to update or acquire the initial lock.
+// Once a proxier is created, it will keep iptables up to date in the background and
+// will not terminate if a particular iptables call fails.
+func NewProxier(ipt utiliptables.Interface) (*Proxier, error) {
+	return &Proxier{
+		serviceMap: make(map[proxy.ServicePortName]*serviceInfo),
+		iptables:   ipt,
+	}, nil
+}
+
+func (proxier *Proxier) sameConfig(info *serviceInfo, service *api.Service, port *api.ServicePort) bool {
+	if info.portal.protocol != port.Protocol || info.portal.port != port.Port {
+		return false
+	}
+	if !info.portal.ip.Equal(net.ParseIP(service.Spec.ClusterIP)) {
+		return false
+	}
+	if !ipsEqual(info.deprecatedPublicIPs, service.Spec.DeprecatedPublicIPs) {
+		return false
+	}
+	if !api.LoadBalancerStatusEqual(&info.loadBalancerStatus, &service.Status.LoadBalancer) {
+		return false
+	}
+	if info.sessionAffinityType != service.Spec.SessionAffinity {
+		return false
+	}
+	return true
+}
+
+func ipsEqual(lhs, rhs []string) bool {
+	if len(lhs) != len(rhs) {
+		return false
+	}
+	for i := range lhs {
+		if lhs[i] != rhs[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// How often we sync iptables
+const syncIntervalIptables = 5 * time.Second
+
+// SyncLoop runs periodic work.  This is expected to run as a goroutine or as the main loop of the app.  It does not return.
+func (proxier *Proxier) SyncLoop() {
+	t := time.NewTicker(syncIntervalIptables)
+	defer t.Stop()
+	for {
+		<-t.C
+		glog.V(6).Infof("Periodic sync")
+		if err := proxier.syncProxyRules(); err != nil {
```

Comment:
putting in a closure with mutex locking next commit.

On Mon, Aug 10, 2015 at 11:42 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36709373>
> :
>
> > +		}
> > +	}
> > +	return true
> > +}
> > +
> > +// How often we sync iptables
> > +const syncIntervalIptables = 5 * time.Second
> > +
> > +// SyncLoop runs periodic work.  This is expected to run as a goroutine or as the main loop of the app.  It does not return.
> > +func (proxier *Proxier) SyncLoop() {
> > +	t := time.NewTicker(syncIntervalIptables)
> > +	defer t.Stop()
> > +	for {
> > +		<-t.C
> > +		glog.V(6).Infof("Periodic sync")
> > +		if err := proxier.syncProxyRules(); err != nil {
>
> This call accesses the serviceMap unsynchronized
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36709373>
> .
>